policy_head.py

```diff
@@ -0,0 +1,187 @@
+from typing import Union, Dict, Optional
+from easydict import EasyDict
+import torch
+import torch.nn as nn
+from copy import deepcopy
+from ding.utils import SequenceType, squeeze, MODEL_REGISTRY
+from ding.model.common import ReparameterizationHead, RegressionHead, DiscreteHead, MultiHead, \
+    FCEncoder, ConvEncoder, IMPALAConvEncoder
+from action import ActionSpace, ActionType
+
+class ActionArgHead(nn.Module):
+    r"""
+    Overview:
+        
+    Interfaces:
+        ``__init__``, ``forward``
+    """
+    def __init__(
+        self,
+        encoded_part_obs_shape,
+        obs_embedding_shape,
+        action_type_logit_shape,
+        # args_shape,
+        hidden_size
+        ):
+        super(ActionArgHead, self).__init__()
+        self.W_k = nn.Linear(encoded_part_obs_shape, hidden_size)
+        self.W_q = nn.Linear(action_type_logit_shape, hidden_size)
+
+    def forward(
+        self,
+        obs_embedding,  # (B, hidden_size)
+        action_type_logit,   # (B, action_type_num)
+        encoded_part_obs     # shape (B, args_shape, encoded_part_obs_shape), part of encoded_obs related to the current arg_head
+        ):
+        # cross attention
+        key = self.W_k(encoded_part_obs)   # (B, args_shape, hidden_size)
+
+        query = self.W_q(action_type_logit) + obs_embedding     # (B, hidden_size)
+        query = query.unsqueeze(1)      # (B, 1, hidden_size)
+
+        logit = torch.mm(query, key.T)  # (B, 1, args_shape)
+        logit = logit.squeeze(1)        # (B, args_shape)
+        return logit
+
+
+class GenshinVAC(nn.Module):
+    r"""
+    Overview:
+        The VAC model for DI-Card.
+    Interfaces:
+        ``__init__``, ``forward``, ``compute_actor``, ``compute_critic``
+    """
+    mode = ['compute_actor', 'compute_critic', 'compute_actor_critic']
+    # action_type_names is a dict corresponding to the sequence number and action type name
+    # e.g {'play_card':0}
+    action_type_names = {getattr(ActionType, attr): attr for attr in dir(ActionType) if not callable(getattr(ActionType, attr)) and not attr.startswith("__")}
+    # action_obs_name_map is used to match action names and encoded_obs names for action_args_head
+    action_obs_name_map = {'play_card':'card_obs','use_skill':'skill_obs', 'change_character': 'character_obs'}
+    def __init__(
+        self,
+        obs_embedding_shape: Union[int, SequenceType],
+        action_shape: Union[int, SequenceType, EasyDict],
+        action_space: ActionSpace,
+        encoded_obs_shape: Dict,    # Should correspond to obs_merge_input_sizes in ObservationEncoder
+        # actor_head_hidden_size: int = 64,
```

Comment:
Because arg_head is three (play_card/use_skill/change_character), each arg_head needs the shape information of the intermediate result encoded_obs of the encoder (of course, the modified encoder now) when initializing, namely obs_merge_input_sizes. I will change it later.